### PR TITLE
issue180 Replace InvalidLRAParticipantDefinitionException with furthe…

### LINKED ISSUE #180: Replace InvalidLRAParticipantDefinitionException with further unspecified deployment/start failure
@tomjenkinson had a good idea that InvalidLRAParticipantDefinitionException may be an unnecessary burden in the future because we don't really care what RTE will be throw. It is only important that it will fail the deployment/start of the application. So we can remove InvalidLRAParticipantDefinitionException from the specification and update the text accordingly.

### PR DESCRIPTION
**Pull request by xstefank**
…r unspecified deployment/start failure

resolves #180 

Signed-off-by: xstefank <xstefank122@gmail.com>

### CODE REVIEW

File: api/src/main/java/org/eclipse/microprofile/lra/annotation/Compensate.java [hunk 1]
```diff
@@ -53,8 +53,8 @@
  *     </code>
  * </pre>
  * would be a valid compensation method declaration. If an invalid signature is detected 
- * the {@link org.eclipse.microprofile.lra.participant.InvalidLRAParticipantDefinitionException} 
- * will be thrown during the application startup.
+ * the implementation of this specification MUST prohibit successful startup of this application
```
Comment (rdebusscher) [inline api/src/main/java/org/eclipse/microprofile/lra/annotation/Compensate.java]: successful startup of **this** application -> successful startup of **the** application

File: api/src/main/java/org/eclipse/microprofile/lra/annotation/Compensate.java [hunk 2]
```diff
@@ -53,8 +53,8 @@
...
+ * (e.g. with the runtime exception).
```
Comment (rdebusscher) [inline api/src/main/java/org/eclipse/microprofile/lra/annotation/Compensate.java]: with **the** runtime exception -> with **a** runtime exception

File: api/src/main/java/org/eclipse/microprofile/lra/annotation/Complete.java [hunk 3]
```diff
@@ -55,8 +55,8 @@
  *     </code>
  * </pre>
  * would be a valid completion method declaration. If an invalid signature is detected 
- * the {@link org.eclipse.microprofile.lra.participant.InvalidLRAParticipantDefinitionException} 
- * will be thrown during the application startup.
+ * the implementation of this specification MUST prohibit successful startup of this application
```
Comment (rdebusscher) [inline api/src/main/java/org/eclipse/microprofile/lra/annotation/Complete.java]: same remark

File: api/src/main/java/org/eclipse/microprofile/lra/annotation/Forget.java [hunk 4]
```diff
@@ -59,8 +59,8 @@
  *     </code>
  * </pre>
  * would be a valid forget method declaration. If an invalid signature is detected 
- * the {@link org.eclipse.microprofile.lra.participant.InvalidLRAParticipantDefinitionException} 
- * will be thrown during the application startup.
+ * the implementation of this specification MUST prohibit successful startup of this application
```
Comment (rdebusscher) [inline api/src/main/java/org/eclipse/microprofile/lra/annotation/Forget.java]: same remark

File: api/src/main/java/org/eclipse/microprofile/lra/annotation/Status.java [hunk 5]
```diff
@@ -57,8 +57,8 @@
  *     </code>
  * </pre>
  * would be a valid status method declaration. If an invalid signature is detected 
- * the {@link org.eclipse.microprofile.lra.participant.InvalidLRAParticipantDefinitionException} 
- * will be thrown during the application startup.
+ * the implementation of this specification MUST prohibit successful startup of this application
```
Comment (rdebusscher) [inline api/src/main/java/org/eclipse/microprofile/lra/annotation/Status.java]: same remark

**Comment by xstefank** [comment]
@rdebusscher Rudy, thanks for the review. Comments addressed in the latest force push.

**Comment by xstefank** [comment]
@mmusgrov TCK is testing that the deployment will fail. This was already agreed upon and merged as part of #148. Why would we allow an application to start if it will fail during the first invocation?

**Comment by ochaloup** [comment]
I think the deployment should be forcibly stopped when does not follow the spec. I think MUST is correct here. I prefer the fail fast practice where it is possible.

**Comment by mmusgrov** [comment]
> @mmusgrov TCK is testing that the deployment will fail. This was already agreed upon and merged as part of #148. Why would we allow an application to start if it will fail during the first invocation?

You are assuming a specific implementation approach for this feature. For example, one implementation might delay checking resource classes until the resource endpoint is accessed so in this case the application would already have deployed.

**Comment by rdebusscher** [comment]
A JAX-RS application fails to deploy when there are 2 resources with the same Path (and not that you have an error when you try to access the URL which is defined twice)

A CDI application fails to deploy when a requested CDI bean doesn't exist (not when you try to use it)

So there are enough other situations where deployment fails when not compliant.

>For example, one implementation might delay checking resource classes until the resource endpoint is accessed so in this case the application would already have deployed

This is what we want to avoid that this is possible. No point of allowing something in production which is known to be failing!

The use of MUST is for the benefit of the developer/user of the spec here.

**Comment by mmusgrov** [comment]
So I was outvoted 3 to 1 so I approved the PR

**Comment by mmusgrov** [comment]
But I can see this being tricky to implement in a pure JAX-RS server.

**Comment by xstefank** [comment]
> But I can see this being tricky to implement in a pure JAX-RS server.

Out of curiosity, I tried with Thorntail, Quarkus and Payara Micro. This fails the deployment in all of them:
```java
@WebListener
public class ContextListener implements ServletContextListener {

    @Override
    public void contextInitialized(ServletContextEvent sce) {
        throw new RuntimeException("ServletContextListener context initialized -- POSSIBLE LRA SCAN");
    }
}
```

**Comment by rdebusscher** [comment]
@xstefank However, pure MicroProfile implementation doesn't need to have Servlets.

@mmusgrov 
>But I can see this being tricky to implement in a pure JAX-RS server.

As mentioned, deployment should fail also when more than 1 JAX-RS resource use the same path. So there is a possibility.
And since it is MicroProfile, you have always CDI available where you can fail the deploy in a CDI extension.

**Comment by mmusgrov** [comment]
> @xstefank However, pure MicroProfile implementation doesn't need to have Servlets.

Section 2.3.2 Servlet of the JAX-RS specification says:

> A JAX-RS application is packaged as a Web application in a .war file.

**Comment by rdebusscher** [comment]
But we aren't allowed to depend on the servlet spec in MicroProfile.

**Comment by xstefank** [comment]
Sorry about that. You are right that is a servlet, not JAX-RS and we cannot rely on it even if it's commonly the base for JAX-RS. 

I agree with Rudy that we should be depending only on existing MicroProfile specifications (umbrella rule). This means that through CDI we can definitely catch the application startup (extensions, `@PostContruct` and `@ApplicationScoped`). However, it seems that in JAX-RS itself doesn't really have a facility to execute code on the application startup. So I will raise this as a new issue and we should decide in the community whether or not we will also require CDI or we will change the wording and the TCK to expect failure on application deployment or on the first request.

However, I would also note that the implementation of LRA will probably always be tied to some container that provides between other bootstrap and implementation of JAX-RS. Cannot we expect that such a container provides a way how modules/libraries can tie into application bootstrap?

**Pull request merged**